File: s2/query.zig
//! query.zig — S2/H3 offline reverse geocoder, Zig runtime (RGEO0001 format).
//!
//! Reads an RGEO0001 binary blob (produced by builder.py) and answers
//! reverse-geocode queries with no heap allocations.
//!
//! The binary uses H3 cells encoded as compact uint32 values in a two-level
//! block binary search structure:
//!   L10 (coarse, res-6):  enc6 = (h3_uint64 >> 27) & 0x1FFFFFF
//!   L12 (fine,   res-7):  enc7 = (h3_uint64 >> 24) & 0xFFFFFFF
//!
//! H3 cell computation (latlng → cell ID) must be performed by the caller.
//! Typical usage from Zig with the h3 C library, or from JavaScript with h3-js.
//!
//! API (library):
//!   const gc = try ReverseGeocoder.init(data_slice);
//!   const admin_id = gc.lookupByEnc(enc6, enc7);  // ?u16, null = ocean
//!
//! WASM exports (compiled with -target wasm32-freestanding):
//!   geocoder_init(ptr, len) i32           — 0 on success, -1 on bad file
//!   geocoder_lookup(enc6, enc7) i32       — admin_id (0–65533) or -1 (ocean)
//!
//! JavaScript caller example:
//!   const cell7 = h3.latLngToCell(lat, lon, 7);
//!   const cell6 = h3.cellToParent(cell7, 6);
//!   const enc7  = encodeRes7(h3.cellToParent(cell7, 6));  // see encoding below
//!   const enc6  = encodeRes6(cell6);
//!   // encodeRes6(cellStr) = ((parseInt(cellStr,16) >> 27n) & 0x1FFFFFFn) | 0
//!   // encodeRes7(cellStr) = ((parseInt(cellStr,16) >> 24n) & 0xFFFFFFFn) | 0

const std = @import("std");

// ---------------------------------------------------------------------------
// Constants (must match builder.py exactly)
// ---------------------------------------------------------------------------

const MAGIC              = "RGEO0001";
const RECORDS_PER_BLOCK: usize = 10;
const RECORD_SIZE:        usize = 6;   // u32 cell_id + u16 admin_id
const BLOCK_SIZE:         usize = 64;
const HEADER_SIZE:        usize = 64;

// ---------------------------------------------------------------------------
// Read helpers (little-endian, unaligned)
// ---------------------------------------------------------------------------

inline fn rdU16(data: []const u8, off: usize) u16 {
    return std.mem.readInt(u16, data[off..][0..2], .little);
}
inline fn rdU32(data: []const u8, off: usize) u32 {
    return std.mem.readInt(u32, data[off..][0..4], .little);
}

// ---------------------------------------------------------------------------
// ReverseGeocoder
// ---------------------------------------------------------------------------

pub const ReverseGeocoder = struct {
    data: []const u8,

    l10_blocks_off:  usize,
    l10_dir_off:     usize,
    l10_block_count: usize,
    l12_blocks_off:  usize,
    l12_dir_off:     usize,
    l12_block_count: usize,

    pub const Error = error{ BadMagic, BadVersion };

    /// Initialise from a raw RGEO0001 binary slice (no allocation).
    pub fn init(data: []const u8) Error!ReverseGeocoder {
        if (data.len < HEADER_SIZE) return Error.BadMagic;
        if (!std.mem.eql(u8, data[0..8], MAGIC)) return Error.BadMagic;
        const version = rdU32(data, 8);
        if (version != 1) return Error.BadVersion;

        // Header layout (see builder.py write_binary_file):
        //  [0:8]  magic       [8:12]  version
        //  [12:16] l10_count  [16:20] l12_count
        //  [20:24] l10_dir_off [24:28] l12_dir_off
        //  [28:32] admin_off  [32:36] name_off
        //  [36:64] padding
        const l10_count    = rdU32(data, 12);
        const l12_count    = rdU32(data, 16);
        const l10_dir_off  = rdU32(data, 20);
        const l12_dir_off  = rdU32(data, 24);

        const l10_block_count = (l10_count + RECORDS_PER_BLOCK - 1) / RECORDS_PER_BLOCK;
        const l12_block_count = (l12_count + RECORDS_PER_BLOCK - 1) / RECORDS_PER_BLOCK;

        // L10 block array: starts right after the 64-byte header
        const l10_blocks_off = HEADER_SIZE;
        // L12 block array: starts right after the L10 directory
        const l12_blocks_off = l10_dir_off + l10_block_count * 4;

        return .{
            .data            = data,
            .l10_blocks_off  = l10_blocks_off,
            .l10_dir_off     = l10_dir_off,
            .l10_block_count = l10_block_count,
            .l12_blocks_off  = l12_blocks_off,
            .l12_dir_off     = l12_dir_off,
            .l12_block_count = l12_block_count,
        };
    }

    /// Reverse-geocode using pre-encoded H3 cell keys.
    ///   enc6 = (h3_cell_uint64 >> 27) & 0x1FFFFFF   (H3 res-6, 25 bits)
    ///   enc7 = (h3_cell_uint64 >> 24) & 0xFFFFFFF   (H3 res-7, 28 bits)
    /// Returns admin_id or null for ocean / unclassified.
    pub fn lookupByEnc(self: *const ReverseGeocoder, enc6: u32, enc7: u32) ?u16 {
        // L10 coarse lookup (res-6 cell)
        if (blockSearch(self.data, enc6, self.l10_blocks_off, self.l10_dir_off, self.l10_block_count)) |id|
            return id;
        // L12 fine lookup (res-7 cell)
        return blockSearch(self.data, enc7, self.l12_blocks_off, self.l12_dir_off, self.l12_block_count);
    }
};

// -- Block binary search (shared by both tables) -----------------------------

fn blockSearch(data: []const u8, cell_id: u32, blocks_off: usize, dir_off: usize, block_count: usize) ?u16 {
    if (block_count == 0) return null;

    // upper_bound: first block with dir[block] > cell_id
    var lo: usize = 0;
    var hi: usize = block_count;
    while (lo < hi) {
        const mid = lo + (hi - lo) / 2;
        if (rdU32(data, dir_off + mid * 4) <= cell_id) {
            lo = mid + 1;
        } else {
            hi = mid;
        }
    }
    if (lo == 0) return null;
    const block_idx = lo - 1;

    const block_start = blocks_off + block_idx * BLOCK_SIZE;
    for (0..RECORDS_PER_BLOCK) |i| {
        const rec_off  = block_start + i * RECORD_SIZE;
        const rec_cell = rdU32(data, rec_off);
        if (rec_cell == 0 and i > 0) break;  // padding sentinel
        if (rec_cell > cell_id) break;
        if (rec_cell == cell_id) return rdU16(data, rec_off + 4);
    }
    return null;
}

// ---------------------------------------------------------------------------
// H3 encoding helpers (mirror of builder.py / s2.js)
// ---------------------------------------------------------------------------

/// Encode an H3 res-6 cell integer to a compact uint32 key.
/// Equivalent to JavaScript: ((BigInt(h3Int) >> 27n) & 0x1FFFFFFn) | 0
pub fn encodeRes6(h3_int: u64) u32 {
    return @truncate((h3_int >> 27) & 0x1FFFFFF);
}

/// Encode an H3 res-7 cell integer to a compact uint32 key.
/// Equivalent to JavaScript: ((BigInt(h3Int) >> 24n) & 0xFFFFFFFn) | 0
pub fn encodeRes7(h3_int: u64) u32 {
    return @truncate((h3_int >> 24) & 0xFFFFFFF);
}

// ---------------------------------------------------------------------------
// WASM / flat-C exports
// ---------------------------------------------------------------------------

var _geocoder: ?ReverseGeocoder = null;

pub export fn geocoder_init(ptr: usize, len: usize) i32 {
    const data: [*]const u8 = @ptrFromInt(ptr);
    _geocoder = ReverseGeocoder.init(data[0..len]) catch return -1;
    return 0;
}

/// Returns admin_id (0–65533) or -1 for ocean / error.
/// enc6 and enc7 must be pre-computed by the JS caller using h3-js.
pub export fn geocoder_lookup(enc6: u32, enc7: u32) i32 {
    const gc = _geocoder orelse return -1;
    return if (gc.lookupByEnc(enc6, enc7)) |id| @intCast(id) else -1;
}

// ---------------------------------------------------------------------------
// Native CLI entry point
// ---------------------------------------------------------------------------

pub fn main() !void {
    const args = try std.process.argsAlloc(std.heap.page_allocator);
    defer std.process.argsFree(std.heap.page_allocator, args);

    if (args.len < 5) {
        std.debug.print("Usage: query <enc6> <enc7> [s2_geo.bin]\n", .{});
        std.debug.print("  enc6/enc7 are pre-encoded H3 cell IDs (decimal integers)\n", .{});
        return;
    }
    const enc6 = try std.fmt.parseInt(u32, args[1], 0);
    const enc7 = try std.fmt.parseInt(u32, args[2], 0);
    const path = if (args.len >= 4) args[3] else "s2_geo.bin";

    const file = try std.fs.cwd().openFile(path, .{});
    defer file.close();
    const data = try file.readToEndAlloc(std.heap.page_allocator, 256 * 1024 * 1024);
    defer std.heap.page_allocator.free(data);

    const gc = try ReverseGeocoder.init(data);
    if (gc.lookupByEnc(enc6, enc7)) |id| {
        std.debug.print("admin_id: {d}\n", .{id});
    } else {
        std.debug.print("ocean / unclassified\n", .{});
    }
}

File: s2/query.py
"""
query.py — Offline Reverse Geocoder Query Engine

Loads the binary file produced by builder.py and answers reverse-geocode
queries in sub-millisecond time using memory-mapped file access and
block binary search.

S2 / H3 note:
  This implementation uses H3 (pip install h3) as a substitute for the
  S2 geometry library.  Cell IDs are encoded as compact uint32 values
  matching the encoding in builder.py:

    encode_res6(h3_int) = (h3_int >> 27) & 0x1FFFFFF   (25 bits, "level 10")
    encode_res7(h3_int) = (h3_int >> 24) & 0xFFFFFFF   (28 bits, "level 12")

  H3 uses an aperture-7 (not quad-tree) hierarchy, so enc6 != enc7 >> 3
  in general.  The res-6 parent of a res-7 cell is obtained via
  h3.cell_to_parent(cell7_str, 6) and then encoded independently.

Query algorithm (mirrors spec §5):
  1. Compute H3 res-7 cell for (lat, lon)  → enc7   (≈ S2 level 12)
  2. Compute H3 res-6 parent of that cell  → enc6   (≈ S2 level 10)
  3. Block binary search L10 (res-6) table:
       a. Binary search directory for enc6  → block index
       b. Linear scan within 64-byte block → admin_id or miss
  4. If hit → lookup admin table, resolve names → return result  (~63% of queries)
  5. If miss → block binary search L12 (res-7) table for enc7
  6. If hit  → lookup admin table, resolve names → return result  (~37% of queries)
  7. If miss in both → return None (ocean or unclassified territory)

Usage as library:
  from query import ReverseGeocoder
  rg = ReverseGeocoder("s2_geo.bin")
  result = rg.lookup(37.7749, -122.4194)
  # {'country': 'United States', 'adm1': 'California', 'adm2': 'San Francisco'}
  rg.close()

  # Context manager:
  with ReverseGeocoder("s2_geo.bin") as rg:
      result = rg.lookup(48.8566, 2.3522)

Usage as CLI:
  python query.py 37.7749 -122.4194
  python query.py 48.8566 2.3522
  python query.py --data /path/to/s2_geo.bin 35.6762 139.6503
"""

import json
import mmap
import os
import struct
import sys
from typing import Dict, Optional

import h3
import zstandard as zstd

try:
    import numpy as np
    _HAS_NUMPY = True
except ImportError:
    _HAS_NUMPY = False

# ── Constants (must match builder.py exactly) ────────────────────────────────

MAGIC             = b"RGEO0001"
H3_RES_COARSE     = 6
H3_RES_FINE       = 7
RECORDS_PER_BLOCK = 10
RECORD_SIZE       = 6    # uint32 cell_id + uint16 admin_id
BLOCK_SIZE        = 64   # bytes — one cache line
ADMIN_ENTRY_SIZE  = 5    # uint8 country_idx + uint16 adm1_idx + uint16 adm2_idx
HEADER_SIZE       = 64   # bytes


# ── Cell ID encoding (must mirror builder.py exactly) ────────────────────────

def _encode_res6(h3_int: int) -> int:
    """Compact uint32 for an H3 res-6 cell (25 bits)."""
    return (h3_int >> 27) & 0x1FFFFFF


def _encode_res7(h3_int: int) -> int:
    """Compact uint32 for an H3 res-7 cell (28 bits)."""
    return (h3_int >> 24) & 0xFFFFFFF


# ── Core lookup engine ───────────────────────────────────────────────────────

class ReverseGeocoder:
    """
    Memory-mapped reverse geocoder.

    The data file is mmap'd read-only; all internal state is immutable after
    __init__.  Concurrent reads from multiple threads are safe.

    Latency (Python, CPython 3.10+, warm file cache):
      Interior cell (L10 hit):    ~1–2 µs
      Boundary cell (L12 hit):    ~1.5–3 µs
      Ocean / miss (both tables): ~2–4 µs
    """

    def __init__(self, data_path: str = "s2_geo.bin") -> None:
        if not os.path.isfile(data_path):
            raise FileNotFoundError(f"Data file not found: {data_path!r}")

        self._file = open(data_path, "rb")
        self._mm   = mmap.mmap(self._file.fileno(), 0, access=mmap.ACCESS_READ)
        self._parse_header()
        self._load_names()
        if _HAS_NUMPY:
            self._build_numpy_tables()

    def close(self) -> None:
        """Release the mmap and file handle."""
        self._mm.close()
        self._file.close()

    def __enter__(self):
        return self

    def __exit__(self, *_):
        self.close()

    # ── Header parsing ───────────────────────────────────────────────────────

    def _parse_header(self) -> None:
        mm = self._mm

        magic = mm[0:8]
        if magic != MAGIC:
            raise ValueError(f"Invalid magic bytes: {magic!r}")

        version,    = struct.unpack_from("<I", mm, 8)
        l10_count,  = struct.unpack_from("<I", mm, 12)
        l12_count,  = struct.unpack_from("<I", mm, 16)
        l10_dir_off,= struct.unpack_from("<I", mm, 20)
        l12_dir_off,= struct.unpack_from("<I", mm, 24)
        admin_off,  = struct.unpack_from("<I", mm, 28)
        name_off,   = struct.unpack_from("<I", mm, 32)

        if version != 1:
            raise ValueError(f"Unsupported file version: {version}")

        self._l10_count    = l10_count
        self._l12_count    = l12_count
        self._l10_dir_off  = l10_dir_off
        self._l12_dir_off  = l12_dir_off
        self._admin_off    = admin_off
        self._name_off     = name_off

        # L10 block array begins immediately after the 64-byte header
        self._l10_blocks_off = HEADER_SIZE

        # Block count = ceil(record_count / RECORDS_PER_BLOCK)
        self._l10_block_count = (l10_count + RECORDS_PER_BLOCK - 1) // RECORDS_PER_BLOCK

        # L12 block array begins right after the L10 directory
        self._l12_blocks_off = l10_dir_off + self._l10_block_count * 4

        self._l12_block_count = (l12_count + RECORDS_PER_BLOCK - 1) // RECORDS_PER_BLOCK

    # ── Name table ───────────────────────────────────────────────────────────

    def _load_names(self) -> None:
        mm = self._mm
        compressed = bytes(mm[self._name_off:])
        dctx = zstd.ZstdDecompressor()
        raw  = dctx.decompress(compressed)
        data = json.loads(raw.decode("utf-8"))
        self._countries: list = data["countries"]
        self._adm1s:     list = data["adm1"]
        self._adm2s:     list = data["adm2"]

    # ── Numpy fast-path tables ───────────────────────────────────────────────

    def _build_numpy_tables(self):
        mm = self._mm
        self._l10_cells_np, self._l10_admins_np = self._extract_flat(
            mm, self._l10_blocks_off, self._l10_block_count, self._l10_count)
        self._l12_cells_np, self._l12_admins_np = self._extract_flat(
            mm, self._l12_blocks_off, self._l12_block_count, self._l12_count)

    def _extract_flat(self, mm, blocks_off, block_count, record_count):
        if block_count == 0 or record_count == 0:
            return np.array([], dtype=np.uint32), np.array([], dtype=np.uint16)
        total = block_count * BLOCK_SIZE
        raw = np.frombuffer(mm[blocks_off:blocks_off + total], dtype=np.uint8)
        blocks = raw.reshape(block_count, BLOCK_SIZE)
        rec = np.ascontiguousarray(blocks[:, :RECORDS_PER_BLOCK * RECORD_SIZE].reshape(-1, RECORD_SIZE))
        cells  = np.frombuffer(np.ascontiguousarray(rec[:, :4]).tobytes(), dtype=np.uint32)
        admins = np.frombuffer(np.ascontiguousarray(rec[:, 4:]).tobytes(), dtype=np.uint16)
        return cells[:record_count].copy(), admins[:record_count].copy()

    # ── Admin table ──────────────────────────────────────────────────────────

    def _lookup_admin(self, admin_id: int) -> Dict[str, str]:
        """Resolve a uint16 admin_id to a {'country', 'adm1', 'adm2'} dict."""
        off = self._admin_off + admin_id * ADMIN_ENTRY_SIZE
        c_idx, a1_idx, a2_idx = struct.unpack_from("<BHH", self._mm, off)
        return {
            "country": self._countries[c_idx],
            "adm1":    self._adm1s[a1_idx],
            "adm2":    self._adm2s[a2_idx],
        }

    # ── Block binary search ──────────────────────────────────────────────────

    def _block_search(
        self,
        cell_id: int,
        blocks_off: int,
        dir_off: int,
        block_count: int,
    ) -> Optional[int]:
        if _HAS_NUMPY and hasattr(self, '_l10_cells_np'):
            # Select which table based on blocks_off
            if blocks_off == self._l10_blocks_off:
                cells, admins = self._l10_cells_np, self._l10_admins_np
            else:
                cells, admins = self._l12_cells_np, self._l12_admins_np
            # Must pass np.uint32 — passing a Python int to a uint32 array
            # triggers Python-level fallback comparisons (250x slower).
            uid = np.uint32(cell_id)
            pos = int(np.searchsorted(cells, uid))
            if pos < len(cells) and cells[pos] == uid:
                return int(admins[pos])
            return None

        # original fallback ...
        if block_count == 0:
            return None
        mm = self._mm
        lo, hi = 0, block_count - 1
        best = -1
        while lo <= hi:
            mid = (lo + hi) >> 1
            first_key, = struct.unpack_from("<I", mm, dir_off + mid * 4)
            if first_key <= cell_id:
                best = mid
                lo = mid + 1
            else:
                hi = mid - 1
        if best < 0:
            return None
        block_start = blocks_off + best * BLOCK_SIZE
        for i in range(RECORDS_PER_BLOCK):
            rec_off = block_start + i * RECORD_SIZE
            rec_cell, = struct.unpack_from("<I", mm, rec_off)
            if rec_cell == 0 and i > 0:
                break
            if rec_cell == cell_id:
                admin_id, = struct.unpack_from("<H", mm, rec_off + 4)
                return admin_id
            if rec_cell > cell_id:
                break
        return None

    # ── Public API ───────────────────────────────────────────────────────────

    def lookup(self, lat: float, lon: float) -> Optional[Dict[str, str]]:
        """
        Reverse-geocode a coordinate.

        Args:
          lat: latitude  in decimal degrees  [-90, +90]
          lon: longitude in decimal degrees  [-180, +180]

        Returns:
          dict with keys 'country', 'adm1', 'adm2' for land areas,
          or None for ocean / unclassified territory.
        """
        # Step 1: compute fine (res-7) cell and encode it
        cell7_str = h3.latlng_to_cell(lat, lon, H3_RES_FINE)
        enc7      = _encode_res7(h3.str_to_int(cell7_str))

        # Step 2: compute coarse (res-6) parent of the fine cell and encode it
        #         (H3 aperture-7 hierarchy: must use cell_to_parent, not bitshift)
        cell6_str = h3.cell_to_parent(cell7_str, H3_RES_COARSE)
        enc6      = _encode_res6(h3.str_to_int(cell6_str))

        # Step 3: L10 lookup (interior cells — ~63% of land queries)
        admin_id = self._block_search(
            enc6,
            self._l10_blocks_off,
            self._l10_dir_off,
            self._l10_block_count,
        )
        if admin_id is not None:
            return self._lookup_admin(admin_id)

        # Step 4: L12 lookup (boundary cells — ~37% of land queries)
        admin_id = self._block_search(
            enc7,
            self._l12_blocks_off,
            self._l12_dir_off,
            self._l12_block_count,
        )
        if admin_id is not None:
            return self._lookup_admin(admin_id)

        # Step 5: not found — ocean or unclassified territory
        return None


# ── CLI entry point ──────────────────────────────────────────────────────────

def main() -> None:
    import argparse

    parser = argparse.ArgumentParser(
        description="Offline reverse geocoder — look up a coordinate in s2_geo.bin.",
        formatter_class=argparse.ArgumentDefaultsHelpFormatter,
    )
    parser.add_argument("lat",  type=float, help="Latitude in decimal degrees.")
    parser.add_argument("lon",  type=float, help="Longitude in decimal degrees.")
    parser.add_argument(
        "--data", "-d",
        default="s2_geo.bin",
        help="Path to the binary data file.",
    )
    args = parser.parse_args()

    with ReverseGeocoder(args.data) as rg:
        result = rg.lookup(args.lat, args.lon)

    print(json.dumps(result, ensure_ascii=False))


if __name__ == "__main__":
    main()
